Add ColorHashTSDFVolume implementation

armanrasta opened the pull request:
# Add ColorHashTSDFVolume implementation  [[#25155](https://github.com/opencv/opencv/issues/25155)]
## Description
Added a new ColorHashTSDFVolume implementation that combines the benefits of HashTSDFVolume's efficient spatial hashing with color support. This provides memory-efficient RGB-D fusion with better performance compared to regular ColorTSDFVolume.

### Key Features
- Hash-based spatial data structure for efficient storage
- Color integration during volume updates
- Raycast with color interpolation
- Compatible with existing TSDF interfaces
- CPU implementation with parallel processing support

### Implementation Details
- Added new ColorHashTSDFVolume class with create() factory method
- ColorVoxel structure combining TSDF and RGB data
- Spatial hashing for efficient voxel lookup
- Weighted running average for color updates
- Trilinear interpolation during raycasting
- Unit tests for basic operations and edge cases

### Files Modified/Added
- modules/3d/src/rgbd/color_hash_volume.hpp - New header defining ColorHashTSDFVolume interface
- modules/3d/src/rgbd/color_hash_volume.cpp - Implementation of ColorHashTSDFVolume
- modules/3d/test/test_color_hash_volume.cpp - Unit tests

### Performance
The implementation uses spatial hashing to only store voxels near surfaces, significantly reducing memory usage compared to regular ColorTSDFVolume while maintaining similar processing speed.

### Testing
Added unit tests that verify:
- Basic integration and raycasting operations
- Empty volume handling
- Memory usage patterns

### Future Work
- GPU/OpenCL implementation
- Additional color interpolation methods
- Extended comparison tests with other volume types

armanrasta commented:
There is an error in the builds, on my test file because i didnt pull a commit here. I'd fix solve it in another pull

savuor reviewed (changes requested):
There are also some minor comments on the code

savuor reviewed (changes requested):
Aside from the other review comments, I'd like to see how it works before approving it. Can you show several pictures of it? There can be raycast results or fetched point clouds with colors. They can be obtained during tests run with `--test_debug` argument.
Please also change the following:
* add `ColorHashTSDF` to `boundingBoxEnableGrowth`, `reproduceVolPoseRot` and `hugeSceneGrowth` tests as a volume type to check
* Docs and comments: mention `ColorHashTSDF` into `volume.hpp` and `volume_settings.hpp` files everywhere where `ColorTSDF` or `HashTSDF` is mentioned

asmorkalov commented:
@armanrasta Friendly reminder.

armanrasta commented:
> @armanrasta Why do you need extra Github actions file?

I wanted to test the action on my own side but made a mistake duplicated an action. Just dropped the commit.

asmorkalov commented:
@armanrasta Thanks for the update. There were several changes in modules structure made. Could you rebase the PR branch and resolve merge issues?

asmorkalov commented:
@savuor Could you take a look on it again.

armanrasta committed:
Add ColorHashTSDF volume to ptcloud module

Implement CPU-only ColorHashTSDF on top of HashTSDF spatial hashing with
ColorTSDF RGB integration. Rebased onto upstream 5.x after modules/3d
was split into ptcloud.

armanrasta committed:
Remove ColorHashTSDF volume from StaticVolumeBoundingBox test instantiation in ptcloud module

armanrasta committed:
Fix trailing whitespace in ptcloud files for CI whitespace check

savuor commented:
The code looks good to me but there are huge visible artifacts on RGB renders produced by the tests:

<img width="640" height="480" alt="0014_render_color" src="https://github.com/user-attachments/assets/dea7f582-36c6-420d-aba0-3cb8bcee3219" />
<img width="640" height="480" alt="0004_render_color" src="https://github.com/user-attachments/assets/b66f3b1a-7222-4cee-b1c5-586d05a92616" />

This is connected to raycast and/or color interpolation code. The same mistake is probably in the original ColorTSDF code.

Looks like I've found a problem, and here's the patch to fix it. Please take a look at it, and at other similar places in other TSDF code, looks like it can be useful there too:

<details><summary>Patch</summary>

```
diff --git a/modules/ptcloud/src/color_hash_tsdf_functions.cpp b/modules/ptcloud/src/color_hash_tsdf_functions.cpp
index 6984665..efae148 100644
--- a/modules/ptcloud/src/color_hash_tsdf_functions.cpp
+++ b/modules/ptcloud/src/color_hash_tsdf_functions.cpp
@@ -482,6 +482,7 @@ void raycastColorHashTsdfVolumeUnit(
                                             std::numeric_limits<int>::min());
                 float tprev = tcurr;
                 float prevTsdf = truncDist;
+                float currTsdf = truncDist;
                 (void)prevVolumeUnitIdx;

                 while (tcurr < tmax)
@@ -491,7 +492,7 @@ void raycastColorHashTsdfVolumeUnit(

                     VolumeUnitIndexes::const_iterator it = volumeUnits.find(currVolumeUnitIdx);

-                    float currTsdf = prevTsdf;
+                    currTsdf = prevTsdf;
                     int currWeight = 0;
                     float stepSize = 0.5f * blockSize;
                     cv::Vec3i volUnitLocalIdx;
@@ -502,35 +503,48 @@ void raycastColorHashTsdfVolumeUnit(
                         volUnitLocalIdx = volumeToVoxelCoord(currRayPos - currVolUnitPos, voxelSizeInv);
                         RGBTsdfVoxel currVoxel = _atColorHash(volUnitsData, volUnitLocalIdx, it->second.index,
                                                               volResolution.x, volDims);
+
                         currTsdf = tsdfToFloat(currVoxel.tsdf);
                         currWeight = currVoxel.weight;
-                        stepSize = tstep;
-                    }

-                    if (prevTsdf > 0.f && currTsdf <= 0.f && currWeight > 0)
-                    {
-                        float tInterp = (tcurr * prevTsdf - tprev * currTsdf) / (prevTsdf - currTsdf);
-                        if (!cvIsNaN(tInterp) && !cvIsInf(tInterp))
+                        if (currTsdf != prevTsdf)
                         {
-                            Point3f pv = orig + tInterp * rayDirV;
-                            Point3f nv = getNormalColorHashVoxel(pv, voxelSizeInv, volumeUnitDegree, volDims,
-                                                                 volUnitsData, volumeUnits);
-                            if (!isNaN(nv))
-                            {
-                                normal = vol2camRot * nv;
-                                point = vol2cam * pv;
-                                if (colors)
-                                    color = getColorHashVoxel(pv, voxelSizeInv, volumeUnitDegree, volDims,
-                                                              volUnitsData, volumeUnits);
-                            }
+                            // from pos to zero or negative
+                            // or from neg to zero or positive
+                            bool posCurr = currTsdf > 0.f, posPrev = prevTsdf > 0.f;
+                            bool negCurr = currTsdf < 0.f, negPrev = prevTsdf < 0.f;
+                            if (posCurr != posPrev || negCurr != negPrev)
+                                break;
                         }
-                        break;
+
+                        stepSize = tstep;
                     }
+
                     prevVolumeUnitIdx = currVolumeUnitIdx;
                     prevTsdf = currTsdf;
                     tprev = tcurr;
                     tcurr += stepSize;
                 }
+
+                if (prevTsdf >= 0.f && currTsdf <= 0.f && prevTsdf > currTsdf)
+                {
+                    float tInterp = (tcurr * prevTsdf - tprev * currTsdf) / (prevTsdf - currTsdf);
+                    if (!cvIsNaN(tInterp) && !cvIsInf(tInterp))
+                    {
+                        Point3f pv = orig + tInterp * rayDirV;
+                        Point3f nv = getNormalColorHashVoxel(pv, voxelSizeInv, volumeUnitDegree, volDims,
+                                                                 volUnitsData, volumeUnits);
+                        if (!isNaN(nv))
+                        {
+                            normal = vol2camRot * nv;
+                            point = vol2cam * pv;
+                            if (colors)
+                                color = getColorHashVoxel(pv, voxelSizeInv, volumeUnitDegree, volDims,
+                                                          volUnitsData, volumeUnits);
+                        }
+                    }
+                }
+
                 ptsRow[x] = toPtype(point);
                 nrmRow[x] = toPtype(normal);
                 if (clrRow)
diff --git a/modules/ptcloud/src/color_tsdf_functions.cpp b/modules/ptcloud/src/color_tsdf_functions.cpp
index 52c92d5..858a4fb 100644
--- a/modules/ptcloud/src/color_tsdf_functions.cpp
+++ b/modules/ptcloud/src/color_tsdf_functions.cpp
@@ -577,7 +577,7 @@ inline float interpolateColor(float tx, float ty, float tz, float vx[8])

 inline v_float32x4 getColorVoxel(const Mat& volume,
     const Vec4i& volDims, const Vec8i& neighbourCoords, const Point3i volResolution,
-    const float voxelSizeInv, const v_float32x4& p)
+    const v_float32x4& p)
 {
     if (v_check_any(v_lt(p, v_float32x4(1.f, 1.f, 1.f, 0.f))) ||
         v_check_any(v_ge(p, v_float32x4((float)(volResolution.x - 2),
@@ -586,14 +586,15 @@ inline v_float32x4 getColorVoxel(const Mat& volume,
         )
         return nanv;

-    v_int32x4 ip = v_floor(p);
+    const v_int32x4 ip = v_floor(p);

     const int xdim = volDims[0], ydim = volDims[1], zdim = volDims[2];
     const RGBTsdfVoxel* volData = volume.ptr<RGBTsdfVoxel>();

-    int ix = v_get0(ip); ip = v_rotate_right<1>(ip);
-    int iy = v_get0(ip); ip = v_rotate_right<1>(ip);
-    int iz = v_get0(ip);
+    v_int32x4 ipr = ip;
+    int ix = v_get0(ipr); ipr = v_rotate_right<1>(ipr);
+    int iy = v_get0(ipr); ipr = v_rotate_right<1>(ipr);
+    int iz = v_get0(ipr);

     int coordBase = ix * xdim + iy * ydim + iz * zdim;
     float CV_DECL_ALIGNED(16) rgb[4];
@@ -607,10 +608,10 @@ inline v_float32x4 getColorVoxel(const Mat& volume,
         b[i] = (float)volData[neighbourCoords[i] + coordBase].b;
     }

-    v_float32x4 vsi(voxelSizeInv, voxelSizeInv, voxelSizeInv, voxelSizeInv);
-    v_float32x4 ptVox = v_mul(p, vsi);
-    v_int32x4 iptVox = v_floor(ptVox);
-    v_float32x4 t = v_sub(ptVox, v_cvt_f32(iptVox));
+    // p is already in voxel-index units (corners were gathered at floor(p)),
+    // so the trilinear weights are just its fractional part. Do NOT rescale by
+    // voxelSizeInv again here — that double-scaling scrambled the weights.
+    v_float32x4 t = v_sub(p, v_cvt_f32(ip));
     float tx = v_get0(t); t = v_rotate_right<1>(t);
     float ty = v_get0(t); t = v_rotate_right<1>(t);
     float tz = v_get0(t);
@@ -630,10 +631,10 @@ inline v_float32x4 getColorVoxel(const Mat& volume,

 inline Point3f getColorVoxel(const Mat& volume,
     const Vec4i& volDims, const Vec8i& neighbourCoords, const Point3i volResolution,
-    const float voxelSizeInv, const Point3f& _p)
+    const Point3f& _p)
 {
     v_float32x4 p(_p.x, _p.y, _p.z, 0.f);
-    v_float32x4 result = getColorVoxel(volume, volDims, neighbourCoords, volResolution, voxelSizeInv, p);
+    v_float32x4 result = getColorVoxel(volume, volDims, neighbourCoords, volResolution, p);
     float CV_DECL_ALIGNED(16) ares[4];
     v_store_aligned(ares, result);
     return Point3f(ares[0], ares[1], ares[2]);
@@ -643,7 +644,7 @@ inline Point3f getColorVoxel(const Mat& volume,
 #else
 inline Point3f getColorVoxel(const Mat& volume,
     const Vec4i& volDims, const Vec8i& neighbourCoords, const Point3i volResolution,
-    const float voxelSizeInv, const Point3f& p)
+    const Point3f& p)
 {
     const int xdim = volDims[0], ydim = volDims[1], zdim = volDims[2];
     const RGBTsdfVoxel* volData = volume.ptr<RGBTsdfVoxel>();
@@ -670,11 +671,11 @@ inline Point3f getColorVoxel(const Mat& volume,
         b[i] = (float)volData[neighbourCoords[i] + coordBase].b;
     }

-    Point3f ptVox = p * voxelSizeInv;
-    Vec3i iptVox(cvFloor(ptVox.x), cvFloor(ptVox.y), cvFloor(ptVox.z));
-    float tx = ptVox.x - iptVox[0];
-    float ty = ptVox.y - iptVox[1];
-    float tz = ptVox.z - iptVox[2];
+    // p is already in voxel-index units (corners gathered at floor(p) above),
+    // so weights are its fractional part — do not rescale by voxelSizeInv again.
+    float tx = p.x - ix;
+    float ty = p.y - iy;
+    float tz = p.z - iz;

     res = Point3f(interpolateColor(tx, ty, tz, r),
         interpolateColor(tx, ty, tz, g),
@@ -868,7 +869,7 @@ void raycastColorTsdfVolumeUnit(const VolumeSettings &settings, const Matx44f &c
                         {
                             v_float32x4 pv = v_add(orig, v_mul(dir, v_setall_f32(ts)));
                             v_float32x4 nv = getNormalColorVoxel(volume, volDims, neighbourCoords, volResolution, pv);
-                            v_float32x4 cv = getColorVoxel(volume, volDims, neighbourCoords, volResolution, voxelSizeInv, pv);
+                            v_float32x4 cv = getColorVoxel(volume, volDims, neighbourCoords, volResolution, pv);

                             if (!isNaN(nv))
                             {
@@ -980,7 +981,7 @@ void raycastColorTsdfVolumeUnit(const VolumeSettings &settings, const Matx44f &c
                         {
                             Point3f pv = (orig + dir * ts);
                             Point3f nv = getNormalColorVoxel(volume, volDims, neighbourCoords, volResolution, pv);
-                            Point3f cv = getColorVoxel(volume, volDims, neighbourCoords, volResolution, voxelSizeInv, pv);
+                            Point3f cv = getColorVoxel(volume, volDims, neighbourCoords, volResolution, pv);
                             if (!isNaN(nv))
                             {
                                 //convert pv and nv to camera space
@@ -1114,7 +1115,7 @@ inline void coord(
                             getNormalColorVoxel(volume, volDims, neighbourCoords, volResolution, p * voxelSizeInv)));
                     if (needColors)
                         colors.push_back(toPtype(pose.rotation() *
-                            getColorVoxel(volume, volDims, neighbourCoords, volResolution, voxelSizeInv, p * voxelSizeInv)));
+                            getColorVoxel(volume, volDims, neighbourCoords, volResolution, p * voxelSizeInv)));
                 }
             }
         }

```

</details>

armanrasta committed:
Fix ColorTSDF color interpolation double-scaling and hash raycast zero-crossing

getColorVoxel was receiving coordinates already in voxel-index units and then
multiplying by voxelSizeInv again, which scrambled trilinear weights and
produced the RGB render artifacts. Drop the extra scale and take the
fractional part of p directly.

Also restructure ColorHashTSDF (and matching HashTSDF) raycast surface
detection: break on TSDF sign change, then interpolate the hit after the
marching loop — same fix as suggested in review.

asmorkalov commented:
@savuor Could you take a look on the current version.

armanrasta commented:
<p class="ui-markdown__paragraph ui-1mdx765 ui-uu7i3w ui-8kmwxx ui-j7cesy ui-13faqbe ui-14l7nz5 ui-zboxd6" style="--tw-space-y-reverse: 0; margin-block: 0px 16px; word-break: break-word; margin-bottom: 6px; margin-top: 6px; color: rgba(228, 228, 228, 0.92); font-family: system-ui, Ubuntu, &quot;Droid Sans&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Hi @savuor,</p><p class="ui-markdown__paragraph ui-1mdx765 ui-uu7i3w ui-8kmwxx ui-j7cesy ui-13faqbe ui-14l7nz5 ui-zboxd6" style="--tw-space-y-reverse: 0; margin-block: 0px 16px; word-break: break-word; margin-bottom: 6px; margin-top: 6px; color: rgba(228, 228, 228, 0.92); font-family: system-ui, Ubuntu, &quot;Droid Sans&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The latest CI run on commit<span> </span><code class="ui-markdown__inline-code ui-rtu7ee ui-1i9v5mb ui-1hhlpkd ui-6lrnw6 ui-115g5cq ui-odjx36 ui-1wd3ewq ui-hayvgc ui-eb7xqv ui-37zpob ui-1lldw8n ui-1mzt3pk" style="border-radius: 5px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); color: rgba(228, 228, 228, 0.92); font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 0.9em; line-height: 1.4; overflow-wrap: break-word; word-break: normal; padding: 1.5px 2px;">77a16a5</code><span> </span>has 4 failing checks, but none of them are related to this PR's scope (<code class="ui-markdown__inline-code ui-rtu7ee ui-1i9v5mb ui-1hhlpkd ui-6lrnw6 ui-115g5cq ui-odjx36 ui-1wd3ewq ui-hayvgc ui-eb7xqv ui-37zpob ui-1lldw8n ui-1mzt3pk md-inline-path-filename-like" style="border-radius: 5px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); color: rgba(228, 228, 228, 0.92); font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 0.9em; line-height: 1.4; overflow-wrap: break-word; word-break: normal; padding: 1.5px 2px;"><span class="ui-markdown__inline-path-prefix ui-1heor9g md-inline-path-prefix" style="color: inherit;">modules/</span><span class="ui-markdown__inline-path-filename ui-1heor9g md-inline-path-filename" style="color: inherit;">ptcloud</span></code><span> </span>/<span> </span><code class="ui-markdown__inline-code ui-rtu7ee ui-1i9v5mb ui-1hhlpkd ui-6lrnw6 ui-115g5cq ui-odjx36 ui-1wd3ewq ui-hayvgc ui-eb7xqv ui-37zpob ui-1lldw8n ui-1mzt3pk" style="border-radius: 5px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); color: rgba(228, 228, 228, 0.92); font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 0.9em; line-height: 1.4; overflow-wrap: break-word; word-break: normal; padding: 1.5px 2px;">ColorHashTSDFVolume</code>):</p><div class="ui-scroll-area ui-1ekbvf6 ui-1logqu4 ui-1fx8caa ui-8ddqcf ui-1n2onr6 ui-6ikm8r ui-10wlt62 ui-rvj5dj ui-n3lufb ui-tr57km ui-14beivq ui-13xjzxd ui-178xt8z ui-13fuv20 ui-1aeic0j ui-s1s249 ui-32b0ac ui-141kqco ui-so031l ui-1q0q8m5 ui-17fyfba ui-e0pwq ui-19ypqd9 ui-1arpupx ui-1043rbw ui-14l7nz5 ui-zboxd6" data-scroll-padding="4" data-visibility="hover" data-direction="horizontal" style="--tw-space-y-reverse: 0; margin-block: 0px 16px; --scroll-area-scroll-padding: 4px; --scrollbar-inset: 1px; --scrollbar-size: 6px; --scrollbar-thumb-top-offset: 6px; grid-template: 1fr / 1fr; border-radius: 6px; display: grid; position: relative; border-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); border-style: solid; border-width: 1px; margin-bottom: 1em; margin-top: 1em; overflow: hidden; color: rgba(228, 228, 228, 0.92); font-family: system-ui, Ubuntu, &quot;Droid Sans&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="ui-scroll-area__viewport ui-15r89dc ui-2lwn1j ui-mz0i5r ui-1pjcqnp ui-1rohswg ui-6ehuon ui-fk6m8 ui-w2csxc ui-10wlt62 ui-k8sdt9 ui-1duton ui-101ixbw ui-exgr1p ui-7p5m3t ui-y5w88m" style="grid-area: 1 / 1; border-radius: inherit; scrollbar-width: none; max-height: 100%; min-height: 0px; overflow: auto hidden; overscroll-behavior: contain auto; scroll-padding: 0px 4px;"><div class="ui-scroll-area__content ui-9f619 ui-gqtt45 ui-193iq5w ui-1us19tq ui-78zum5 ui-14atkfc ui-g7h5cd ui-1q0g3np" style="box-sizing: border-box; display: flex; flex-direction: row; height: fit-content; max-width: 100%; min-height: 100%; min-width: 100%; width: auto;">
Check | Failing test | Module
-- | -- | --
Linux / Ubuntu (26.04) | opencv_test_calib — killed by 20 min timeout (-9) | calib
Windows / x64 | opencv_perf_geometry → PointsNum_Algo_solvePnP.solvePnP/4 | geometry
Windows / x86 | opencv_test_dnn + opencv_test_dnn_classic → DNNTestNetwork.AlexNet/0 | dnn
macOS-ARM64 | opencv_test_dnn (YuNet, ResNet50_QDQ_ONNX) + opencv_test_features (Features2d_DISK ×5) | dnn, features2d

</div></div><div class="ui-scroll-area__scrollbar ui-10l6tqk ui-1n327nk ui-67bb7w ui-1lliihq ui-y36ywk ui-80663w ui-zkb8sy ui-1gncbth ui-17roi93 ui-14atkfc ui-a38mao" data-orientation="horizontal" data-scrollable="true" role="presentation" aria-hidden="true" style="display: block; pointer-events: auto; position: absolute; z-index: 10; inset: auto 1px 1px; height: 6px; width: auto;"><div class="ui-10l6tqk ui-1i4c3av ui-1m4ooaa ui-1kik0b9 ui-19p5jqy ui-gk890x ui-67bb7w ui-oww4n8 ui-1e8aii2 ui-1vaoic9 ui-144st5w ui-1aiz5cx ui-13vifvy ui-wukr4l ui-1ey2m1c ui-u96u03 ui-14atkfc ui-a38mao ui-65nank ui-cbmsab ui-g01cxk ui-4gyw5p ui-gorybd" data-orientation="horizontal" style="border-radius: 9999px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); opacity: 0; pointer-events: auto; position: absolute; transform: translateX(0px); transition-duration: 0.15s, 0.15s; transition-property: background-color, opacity; transition-timing-function: ease, ease; inset: 0px auto 0px 0px; height: 6px; min-height: auto; min-width: 20px; width: 266px;"></div></div></div><p class="ui-markdown__paragraph ui-1mdx765 ui-uu7i3w ui-8kmwxx ui-j7cesy ui-13faqbe ui-14l7nz5 ui-zboxd6" style="--tw-space-y-reverse: 0; margin-block: 0px 16px; word-break: break-word; margin-bottom: 6px; margin-top: 6px; color: rgba(228, 228, 228, 0.92); font-family: system-ui, Ubuntu, &quot;Droid Sans&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">For reference, the<span> </span><code class="ui-markdown__inline-code ui-rtu7ee ui-1i9v5mb ui-1hhlpkd ui-6lrnw6 ui-115g5cq ui-odjx36 ui-1wd3ewq ui-hayvgc ui-eb7xqv ui-37zpob ui-1lldw8n ui-1mzt3pk" style="border-radius: 5px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); color: rgba(228, 228, 228, 0.92); font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 0.9em; line-height: 1.4; overflow-wrap: break-word; word-break: normal; padding: 1.5px 2px;">ptcloud</code><span> </span>module (where this PR's changes live) builds and passes on every runner, e.g. on macOS-ARM64:</p><div class="composer-message-codeblock" style="--tw-space-y-reverse: 0; margin-block: 0px 16px; margin-bottom: 8px; margin-top: 8px; color: rgba(228, 228, 228, 0.92); font-family: system-ui, Ubuntu, &quot;Droid Sans&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="ui-code-block ui-1n2onr6 ui-1om1abp ui-14l7nz5 ui-1yf7rl7 ui-1qaq26d ui-zboxd6 ui-j3b58b ui-1vb4fzf ui-14gibnr ui-10a2chs ui-1u51yxl ui-178xt8z ui-13fuv20 ui-1aeic0j ui-s1s249 ui-32b0ac ui-141kqco ui-so031l ui-1q0q8m5 ui-17fyfba ui-e0pwq ui-19ypqd9 ui-1arpupx ui-6ikm8r ui-10wlt62 ui-1ua6jya ui-1k89b3v ui-8nuxf3 ui-w0qwo3 ui-y3gzh4 ui-1gpzi3k" style="--ui-code-block-copy-overlay-display: none; background-color: rgb(24, 24, 24); position: relative; border-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); border-radius: 8px; border-style: solid; border-width: 1px; margin: 0px; overflow: hidden; min-height: 66px;"><div class="ui-code-block-content ui-1n2onr6" style="position: relative; min-height: 66px;"><div class="ui-scroll-area ui-1ekbvf6 ui-1logqu4 ui-1fx8caa ui-8ddqcf ui-1n2onr6 ui-6ikm8r ui-10wlt62 ui-rvj5dj ui-n3lufb" data-scroll-padding="4" data-visibility="hover" data-direction="horizontal" style="--scroll-area-scroll-padding: 4px; --scrollbar-inset: 1px; --scrollbar-size: 6px; --scrollbar-thumb-top-offset: 6px; grid-template: 1fr / 1fr; display: grid; position: relative; overflow: hidden;"><div class="ui-scroll-area__viewport ui-15r89dc ui-2lwn1j ui-mz0i5r ui-1pjcqnp ui-1rohswg ui-6ehuon ui-fk6m8 ui-w2csxc ui-10wlt62 ui-k8sdt9 ui-1duton ui-101ixbw ui-exgr1p ui-7p5m3t ui-y5w88m" style="grid-area: 1 / 1; border-radius: inherit; scrollbar-width: none; max-height: 100%; min-height: 0px; overflow: auto hidden; overscroll-behavior: contain auto; scroll-padding: 0px 4px;"><div class="ui-scroll-area__content ui-9f619 ui-gqtt45 ui-193iq5w ui-1us19tq ui-78zum5 ui-14atkfc ui-g7h5cd ui-1q0g3np" style="box-sizing: border-box; display: flex; flex-direction: row; height: fit-content; max-width: none; min-height: 100%; min-width: 100%; width: max-content;"><div class="ui-default-code ui-1wm8ruf ui-1lfj9qp ui-1yxxptd ui-1ua6jya ui-1m21r4p ui-cj37ws ui-13r5tfd ui-uve7l6 ui-1xpa7k ui-1rgtt3y ui-1uhho1l ui-o9w7m5 ui-17ewauw ui-1vakzcb ui-105xmc4 ui-1nsav61 ui-1jxofbz ui-947h55 ui-ta7kcf ui-1iagb1a ui-ez8t0j ui-i2wo5x ui-1iq2d2k ui-1qoclnz ui-k609h7 ui-1rqnq8e ui-19gwdp1 ui-1yfyyyg ui-1apie5d ui-dcvkk1 ui-code-block-default-code" style="--shiki-background: transparent; --shiki-foreground: rgba(228, 228, 228, 0.92); --shiki-token-attribute: #AAA0FA; --shiki-token-class: #EFB080; --shiki-token-comment: #E4E4E45E; --shiki-token-constant-variable: #AAA0FA; --shiki-token-constant: #f8c762; --shiki-token-function: #EFB080; --shiki-token-keyword: #82D2CE; --shiki-token-language-variable: #CC7C8A; --shiki-token-link: #87c3ff; --shiki-token-parameter: #D6D6DD; --shiki-token-property: #AAA0FA; --shiki-token-punctuation: rgba(228, 228, 228, 0.92); --shiki-token-string-expression: #E394DC; --shiki-token-string: #E394DC; --shiki-token-tag: #82D2CE; --shiki-token-type: #EFB080; --shiki-token-variable: #87C3FF; background-color: transparent; contain: layout style paint; font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 12px; font-weight: 400; line-height: 20px; tab-size: 4; padding: 6px 0px; --ui-default-code-tab-size: 4; min-height: 0px; width: max-content; min-width: 100%;"><div class="ui-1pshirs ui-awy1a7" style="contain: layout style; min-width: fit-content;"><div class="ui-78zum5 ui-1f4zecs" style="display: flex; min-height: 20px;"><div class="ui-default-code__line-content ui-98rzlu ui-1sdyfia ui-b3r6kr ui-lyipyv ui-1lfpczk ui-lgbajd ui-1yxiud8 ui-1n2onr6" style="flex: 1 1 0%; overflow: hidden; color: rgba(228, 228, 228, 0.92); position: relative; text-overflow: ellipsis; white-space: pre; padding-left: 6px; padding-right: 8px;"><span>[==========] Running 627 tests from 24 test cases.</span></div></div><div class="ui-78zum5 ui-1f4zecs" style="display: flex; min-height: 20px;"><div class="ui-default-code__line-content ui-98rzlu ui-1sdyfia ui-b3r6kr ui-lyipyv ui-1lfpczk ui-lgbajd ui-1yxiud8 ui-1n2onr6" style="flex: 1 1 0%; overflow: hidden; color: rgba(228, 228, 228, 0.92); position: relative; text-overflow: ellipsis; white-space: pre; padding-left: 6px; padding-right: 8px;"><span>[==========] 627 tests from 24 test cases ran. (75886 ms total)</span></div></div><div class="ui-78zum5 ui-1f4zecs" style="display: flex; min-height: 20px;"><div class="ui-default-code__line-content ui-98rzlu ui-1sdyfia ui-b3r6kr ui-lyipyv ui-1lfpczk ui-lgbajd ui-1yxiud8 ui-1n2onr6" style="flex: 1 1 0%; overflow: hidden; color: rgba(228, 228, 228, 0.92); position: relative; text-overflow: ellipsis; white-space: pre; padding-left: 6px; padding-right: 8px;"><span>[  PASSED  ] 627 tests.</span></div></div></div></div></div></div><div class="ui-scroll-area__scrollbar ui-10l6tqk ui-1n327nk ui-67bb7w ui-1lliihq ui-y36ywk ui-80663w ui-zkb8sy ui-1gncbth ui-17roi93 ui-14atkfc ui-a38mao" data-orientation="horizontal" data-scrollable="true" role="presentation" aria-hidden="true" style="display: block; pointer-events: auto; position: absolute; z-index: 10; inset: auto 1px 1px; height: 6px; width: auto;"><div class="ui-10l6tqk ui-1i4c3av ui-1m4ooaa ui-1kik0b9 ui-19p5jqy ui-gk890x ui-67bb7w ui-oww4n8 ui-1e8aii2 ui-1vaoic9 ui-144st5w ui-1aiz5cx ui-13vifvy ui-wukr4l ui-1ey2m1c ui-u96u03 ui-14atkfc ui-a38mao ui-65nank ui-cbmsab ui-g01cxk ui-4gyw5p ui-gorybd" data-orientation="horizontal" style="border-radius: 9999px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); opacity: 0; pointer-events: auto; position: absolute; transform: translateX(0px); transition-duration: 0.15s, 0.15s; transition-property: background-color, opacity; transition-timing-function: ease, ease; inset: 0px auto 0px 0px; height: 6px; min-height: auto; min-width: 20px; width: 220px;"></div></div></div></div></div></div><p class="ui-markdown__paragraph ui-1mdx765 ui-uu7i3w ui-8kmwxx ui-j7cesy ui-13faqbe ui-14l7nz5 ui-zboxd6" style="--tw-space-y-reverse: 0; margin-block: 0px 16px; word-break: break-word; margin-bottom: 6px; margin-top: 6px; color: rgba(228, 228, 228, 0.92); font-family: system-ui, Ubuntu, &quot;Droid Sans&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">and<span> </span><code class="ui-markdown__inline-code ui-rtu7ee ui-1i9v5mb ui-1hhlpkd ui-6lrnw6 ui-115g5cq ui-odjx36 ui-1wd3ewq ui-hayvgc ui-eb7xqv ui-37zpob ui-1lldw8n ui-1mzt3pk md-inline-variable-like" style="border-radius: 5px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); color: rgba(228, 228, 228, 0.92); font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 0.9em; line-height: 1.4; overflow-wrap: break-word; word-break: normal; padding: 1.5px 2px;">opencv_perf_ptcloud</code><span> </span>reports<span> </span><code class="ui-markdown__inline-code ui-rtu7ee ui-1i9v5mb ui-1hhlpkd ui-6lrnw6 ui-115g5cq ui-odjx36 ui-1wd3ewq ui-hayvgc ui-eb7xqv ui-37zpob ui-1lldw8n ui-1mzt3pk" style="border-radius: 5px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); color: rgba(228, 228, 228, 0.92); font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 0.9em; line-height: 1.4; overflow-wrap: break-word; word-break: normal; padding: 1.5px 2px;">Success</code><span> </span>on Linux 26.04. The<span> </span><code class="ui-markdown__inline-code ui-rtu7ee ui-1i9v5mb ui-1hhlpkd ui-6lrnw6 ui-115g5cq ui-odjx36 ui-1wd3ewq ui-hayvgc ui-eb7xqv ui-37zpob ui-1lldw8n ui-1mzt3pk" style="border-radius: 5px; background-color: color(srgb 0.894118 0.894118 0.894118 / 0.0737255); color: rgba(228, 228, 228, 0.92); font-family: &quot;Droid Sans Mono&quot;, &quot;monospace&quot;, monospace; font-size: 0.9em; line-height: 1.4; overflow-wrap: break-word; word-break: normal; padding: 1.5px 2px;">docs / BuildAndTest</code><span> </span>and all other Linux/Android/macOS-x64/iOS checks are green too.</p><p class="ui-markdown__paragraph ui-1mdx765 ui-uu7i3w ui-8kmwxx ui-j7cesy ui-13faqbe ui-14l7nz5 ui-zboxd6" style="--tw-space-y-reverse: 0; margin-block: 0px 16px; word-break: break-word; margin-bottom: 6px; margin-top: 6px; color: rgba(228, 228, 228, 0.92); font-family: system-ui, Ubuntu, &quot;Droid Sans&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">These 4 failures look like pre-existing upstream flakes (calib timeout, DNN model accuracy/download flakes, solvePnP numerical flake on Windows). Could someone with permissions re-run these jobs, or confirm they're known flakes?
